text/3845-repr-ordered-fields.md

```diff
@@ -0,0 +1,326 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* This will cause a large amount of churn in the Rust ecosystem
+* If we don't end up switching `repr(C)` to mean the system layout/ABI, then we will have two identical reprs, which may cause confusion.
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* `crabi`: http://github.com/rust-lang/rfcs/pull/3470
+    * Currently stuck in limbo since it has a much larger scope. doesn't actually serve to give a consistent cross-platform layout, since it defers to `repr(C)` (and it must, for its stated goals)
+* https://internals.rust-lang.org/t/consistent-ordering-of-struct-fileds-across-all-layout-compatible-generics/23247
+    * This doesn't give a predictable layout that can be used to match the layouts (or prefixes) of different structs
+* https://github.com/rust-lang/rfcs/pull/3718
+    * This one is currently stuck due to a larger scope than this RFC
+* do nothing
+    * We keep getting bug reports on Windows (and other platforms), where `repr(C)` doesn't actually match the target C compiler, or we break a bunch of subtle unsafe code to match the target C compiler.
+# Prior art
+[prior-art]: #prior-art
+
+See Rationale and Alternatives as well
+
+* https://rust-lang.zulipchat.com/#narrow/channel/213817-t-lang/topic/expand.2Frevise.20repr.28.7BC.2Clinear.2C.2E.2E.2E.7D.29.20for.202024.20edition
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* The migration plan, as a whole, needs to be ironed out
+    * Currently, it is just a sketch, but we need timelines, dates, and guarantees to switch `repr(C)` to match the layout algorithm of the target C compiler.
+    * Before this RFC is accepted, t-compiler will need to commit to fixing the layout algorithm sometime in the next edition.
+* The name of the new repr `repr(ordered_fields)` is a mouthful (intentionally for this RFC), maybe we could pick a better name? This could be done after the RFC is accepted.
+    * `repr(linear)`
+    * `repr(ordered)`
+    * `repr(sequential)`
+    * `repr(consistent)`
+    * something else?
+* Is the ABI of `repr(ordered_fields)` specified (making it safe for FFI)? Or not?
+* Should unions expose some niches?
+    * For example, if all variants of the union are structs which have a common prefix, then any niches of that common prefix could be exposed (i.e. in the enum case, making union of structs behave more like an enum).
+    * This must be answered before stabilization, as it is set in stone after that
+* Should this `repr` be versioned?
+    * This way we can evolve the repr (for example, by adding new niches)
+* Should we change the meaning of `repr(C)` in editions <= 2024 after we have reached edition 2033? Yes, it's a breaking change, but at that point it will likely only be breaking code no one uses.
+    * Leaning towards no
+* Should we warn on `repr(ordered_fields)` when explicit tag type is specified (i.e. no `repr(u8)`/`repr(i32)`)
```

Comment:
I think we should not warn when there's an explicit tag type, since there are common valid reasons to have an explicit tag type, you're trying to get a specific layout rather than defaulting to whatever tag type rustc feels like picking today. e.g. having a struct that matches some specific file format:
```rust
#[repr(u32)]
pub enum WavFmtSize {
    Sixteen = 0x10u16.to_le(),
}

#[repr(ordered_fields, u32)]
pub enum WavFileChunk {
    Fmt {
        fmt_size: WavFmtSize,
        ...
    } = u32::from_ne_bytes(b"fmt "),
    DataHeader {
        data_size: U32LE,
    } = u32::from_ne_bytes(b"data"),
}
```

Reply:
Oops, I meant if an explicit tag type is missing